Improve CC documentation about CPU capacity and utilization (#5992)
strimzi/strimzi-kafka-operator: documentation/modules/cruise-control/ref-cruise-control-configuration.adoc

Because Strimzi Kafka brokers are homogeneous, Cruise Control applies the same capacity limits to every broker it is monitoring.

[NOTE]
====
We recommend that the Java memory of the Kafka brokers fits into the `Kafka.spec.kafka.resources.requests.memory`, so that brokers will have roughly the same capacity.
====

> 

Reviewer: Sorry, maybe my comment should have been clearer last time as it could have been read as suggesting to fix the text. Apologies for that. This is more or less correct now, but how does it fit here? Nothing in here talks about memory management or JVM memory configuration. 

Also, in general, the fact that all the brokers have the same capacity is a limitation of Strimzi and not its goal. The test seems to suggest it is the end-goal. The main reason to have the Java memory fit into the memory request is also not because of capacity planning but to achieve stability and avoid running out of memory.

TBH, I would just remove this part and leave this part of the docs as it was.

Author: Yeah, I think there was some confusion due to multiple edits. 

What I wanted to note here is that CC assumes same CPU/memory capacity on all brokers. It simply does not support per-broker values yet. In order to have a stable RebalanceProposal, which is still valid few minutes later when it is applied, it would be good to set requests == limits in Kafka CR, so that resources would be reserved upfront.

Reviewer: @fvaleri Ok, I guess that got lost somewhere between the different review comments. I think it is ok to cover that in that case. But it should not be presented as recommendation because it is really a limitation. So maybe it could say something like `CPU goals in Cruise Control currently don't support CPU limits. For properly working CPU goals, you should always use the same CPU limit ad CPU request in the Kafka brokers configuration` (doesn't need to be exactly this of course). That makes it clear it is not a general recommendation but rather a limitation of Cruise Control. Also - I assume it is only the CPU goals which care about it - so it makes it clear that you can use CPU limits and just use other goals.

@kyguy @tomncooper Is it the case that Cruise Control doesn't handle CPU limits? If yes, what are the plans to fix it? It sounds like quite significant limitation to me.

Reviewer 2: Cruise Control cannot accurately compare CPU utilization between broker pods on different nodes with heterogeneous CPU cores without either:
(1) Broker pod CPU limit being set to a homogenous value through K8s `LimitRange` [1] object or  `Kafka` resource limits
(2) Node specific CPU information in the `spec.cruiseControl.capacity` config in `Kafka` resource. 

Currently, Strimzi does not support (2) and therefore must rely on (1) for accurate CPU utilization comparison between brokers. I feel this follows Kubernetes principles of abstracting physical resources from the applications and users. Otherwise, we would need Strimzi or Cruise Control to be able to gather node information to generate (2) or have users manually track and update (2) whenever a broker pod was scheduled to a different node. 

[1] https://kubernetes.io/docs/concepts/policy/limit-range/

Author: Thanks @kyguy. So this is actually a Strimzi limitation at the moment.

If you agree, I would phrase that note like this:

"Strimzi does not allow to configure per-broker CPU capacity. For that reason, Cruise Control cannot accurately compare CPU utilization between broker pods on different Kubernetes nodes. To overcome this, we suggest to set CPU limits equal to requests in `Kafka.spec.kafka.resources`. This would allow to reserve the required CPU resources upfront and generate accurate RebalanceProposals."

Then, there is also the workaround of excluding CPU goals from the preset hard/default goals, but the configuration is verbose and I don't think we should even mention it in the documentation.

Reviewer: * You cannot phrase some Cruise Control specific limitation as general recommendation. It is not really recommendation, it is a workaround for some limitation. It needs to be phrased as such - so we do not suggest it to anyone, but it is a workaround people can use if they want.
* _Strimzi does not allow to configure per-broker CPU capacity._ I'm not sure how this is related. Per-broker CPU configuration has nothing to do with requests versus limits.
* As far as I understood, it affects Cruise Control only when CPU goals are used. That needs to be clear from it as well.
* _So this is actually a Strimzi limitation at the moment._ ... I do not see how this is a Strimzi limitation. The CPU limit is  something what needs to be understood by Cruise Control. They it can be either set in the CR or it can be detected by the CC metrics reporter. That is my understanding of the issue.

Author: CC allows to set CPU capacity for each broker:
https://github.com/linkedin/cruise-control/blob/migrate_to_kafka_2_4/config/capacityCores.json

Strimzi does not allow to use this type of configuration, as you cannot specify a different CPU limit for each broker. We have the concept of a single "virtual" CPU core that has the cycles of `Kafka.spec.kafka.resources.limits.cpu` cores. 

If you don't set requests == limits. The real/current CPU capacity would be different when broker pods are running on different Kubernetes nodes. It may be due to other workloads scheduled on the same nodes or simply because they are over committed. This makes the RebalanceProposals unreliable with regards to CPU goals (enabled by default).

As an enhancement, we may think to detect the real/current CPU capacity for each broker at runtime (e.g. using metrics as you suggests) and reconcile the CC capacity configuration periodically.

Do we agree with that?

Reviewer: But this is not about setting CPU cores per broker. Strimzi currently doesn't allow this anyway, although it will probably in the future. The issue is that when you set the limit, the actual CPU available can change every second. You configure the resources as request=4cpu and limit=8cpu. That means that in one second, it might have 8cpu, next it might have 6cpu and the next one it might have 4cpu. So being able to set the cpu per-broker in some Cruise Control settings will not help you in anyway because the actual usage might change every second.

Author: The point is: if you set requests == limits in Kafka CR, then the CPU capacity would be guaranteed and rebalance proposal accurate. This is what I would like to recommend when using CC with CPU goals enabled.

Reviewer: Yes, but as I said, there are many possible pros and cons of having requests == limits. So I don't think we should recommend it. It should be just described as workaround.

Reviewer 2: >  It is not really recommendation, it is a workaround for some limitation. It needs to be phrased as such - so we do not suggest it to anyone, but it is a workaround people can use if they want.

Although I originally wanted this as a recommendation, I understand where Jakub is coming from here. The limit setting is only needed for CC to work in specific circumstances (it is not needed for Strimzi as a whole) so I see how characterizing the limit setting as a workaround would make sense.

> The CPU limit is something what needs to be understood by Cruise Control. They it can be either set in the CR or it can be detected by the CC metrics reporter. That is my understanding of the issue.

As discussed above, we cannot set it in the CR at the moment but I don't think we should allow it anyway because it would introduce a user maintenance burden. It would be nice if the CC metrics reporter detected and reported broker capacity and I am not sure why it doesn't  but I am sure there is a good reason. I'll find out why

> Is it the case that Cruise Control doesn't handle CPU limits? If yes, what are the plans to fix it? It sounds like quite significant limitation to me.

The advertised approach to solve this issue without limit setting is to write our own BrokerCapacityConfigResolver [1] I planned to investigate this anyway this to address related network I/O capacity issues, we can look into solving the CPU capacity issue at the same time. 

As for this PR, I am fine with stating limit setting as a workaround for accurate CPU balances when pods are running on nodes with heterogeneous CPU resources.

[1] https://github.com/linkedin/cruise-control/wiki/Pluggable-Components#broker-capacity-config-resolver

Author: > As for this PR, I am fine with stating limit setting as a workaround for accurate CPU balances when pods are running on nodes with heterogeneous CPU resources.

@kyguy if you set `limits=3 < requests=1` in Kafka CR, then you are still not guaranteed to have accurate CPU balances, as each broker real CPU capacity would be somewhere between 1 and 3 cores. The configured limit of 3 cores could not be satisfied for long time, for example because the K8s node does not have capacity. 

Am I wrong?

Reviewer 2: Sorry @fvaleri what I meant to say I am fine with your workaround of setting of `requests==limits`, you are correct!

> We recommend that `Kafka.spec.kafka.resources.requests` and `Kafka.spec.kafka.resources.limits` are set to equal values. With these settings, broker pods will have the same capacity and quality of service. The settings can improve stability. This is especially important when using nodes with heterogeneous resources, such as CPU cores.

Reviewer 3: ```suggestion
We recommend that `Kafka.spec.kafka.resources.requests` and `Kafka.spec.kafka.resources.limits` are set to equal values. With these settings, broker pods will have the same capacity and quality of service is guaranteed. The settings can improve stability. This is especially important when using nodes with heterogeneous resources, such as CPU cores.
```

> [NOTE]
> ====
> In order to guarantee accurate RebalanceProposals when using CPU goals, you can set requests equal to limits in Kafka CR. That way, all CPU resources are reserved upfront and the broker's full capacity is always available, even when the Kubernetes node is under pressure. This is a tradeoff, as it comes at the cost of some unused CPU cycles.
> ====
> 

Reviewer: ```suggestion
[NOTE]
====
In order to guarantee accurate rebalance proposal when using CPU goals, you can set CPU requests equal to CPU limits in `spec.kafka.resources` in the Kafka CR. 
That way, all CPU resources are reserved upfront and are always available.
That allows Cruise Control to properly evaluate the CPU utilization when preparing the rebalance proposals based on CPU goals.
====

```

> Cruise Control assumes brokers resources to be homogeneous. Therefore, it is recommended to set both `Kafka.spec.kafka.resources.limits.cpu` and `Kafka.spec.kafka.resources.limits.memory`, so that Kubernetes will ensure broker pods are limited to the same capacity. This is especially important when using nodes with heterogeneous resources (e.g. CPU cores).

Reviewer: We for sure do not recommend setting memory limits much -> it is not guaranteed and can cause problems when the memory needs to be returned.

Author: Yeah, I mean, the full recommendation would be to set requests == limits. Should I add this point too?

> 

Reviewer: I don't think we recommend this. The only thing we recommend is to make sure that the Java memory of the Kafka brokers fits into the memory request.

Author: OK let me rephrase it.

> Cruise Control assumes brokers resources to be homogeneous. Therefore, it is recommended to set value in `Kafka.spec.kafka.resources.requests` equal to `Kafka.spec.kafka.resources.limits`, so that Kubernetes will ensure broker pods have the same capacity and Guaranteed QoS, which also improves stability. This is especially important when using nodes with heterogeneous resources (CPU cores for example).

Reviewer 3: ```suggestion
We recommend that `Kafka.spec.kafka.resources.requests` and `Kafka.spec.kafka.resources.limits` are set to equal values. With these settings, broker pods will have the same capacity and quality of service. The settings can improve stability. This is especially important when using nodes with heterogeneous resources, such as CPU cores.
```

Reviewer 3: Don't think we need the homogeneous line,as it's written in the sentence before the note.

Author: Perfect. Thanks.

> Cruise Control assumes brokers resources to be homogeneous. Therefore, it is recommended to set value in `Kafka.spec.kafka.resources.requests` equal to `Kafka.spec.kafka.resources.limits`, so that Kubernetes will ensure broker pods have the same capacity and Guaranteed QoS, which also improves stability. This is especially important when using nodes with heterogeneous resources (CPU cores for example).

Reviewer 2: "Cruise Control" -> "Strimzi Cruise Control"

It is recommended to set `Kafka.spec.kafka.resources.limits.cpu` so Kubernetes will ensure broker pods have the same CPU capacity, otherwise the CPU utilization for brokers can get under or overestimated when working with nodes with heterogeneous CPU resources. 

I need to think more about the setting of `Kafka.spec.kafka.resources.limits.cpu`. From what I understand, Cruise Control doesn't have any explicit protections against heterogeneous memory capacities but memory is considered when calculating broker load by using certain metrics (e.g. by using metrics like bytes in and bytes out in its CPU and Network utilization calculations). Values like bytes in and bytes out could differ between brokers with different memory capacities. I am not sure how big of an impact it would have or if it is covered by the other metrics but I would think it would be a good idea to set `Kafka.spec.kafka.resources.limits.cpu` (except for the caveats that Jakub mentioned!)

  name: my-cluster
spec:
  # ...
  kafka:
    # ...
    resources:

>       limits:

Reviewer: Should this be request rather than limit?

Author: This sets requests == limits.

Reviewer: Yes, but it is not clear to everyone. Setting only request also means that it will be the limit but it is much more clear what it means to everyone.

Author: Ok.
